### PR TITLE
universal-android-debloater: unstable-2023-04-10

### PR DESCRIPTION
**Pull request by gesperon**
## Description of changes

New package universal-android-debloater.

## Things done

EDIT:
removed maintainer-list changes from this PR and created https://github.com/NixOS/nixpkgs/pull/267742

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/misc/universal-android-debloater/default.nix [hunk 1]
```diff
@@ -0,0 +1,59 @@
+{ pkgs ? import <nixpkgs> { system = builtins.currentSystem; }
```
Comment (mdarocha) [inline pkgs/misc/universal-android-debloater/default.nix:1]: This should not be done in `nixpkgs`. Instead, every required package should be an argument here. That way, they can be overridden if necessary

File: pkgs/misc/universal-android-debloater/default.nix [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  doCheck = false;
+
+  buildFeatures = [ "iced/glow" ];
+  postInstall = with pkgs; ''
+    wrapProgram $out/bin/uad_gui \
+      --prefix LD_LIBRARY_PATH : ${lib.makeLibraryPath [ fontconfig freetype libglvnd xorg.libX11 xorg.libXcursor xorg.libXrandr xorg.libXi ]} \
+      --suffix PATH : ${lib.makeBinPath [ android-tools ]}
+    '';
+
+  meta = with lib; {
```
Comment (mdarocha) [inline pkgs/misc/universal-android-debloater/default.nix:53]: You don't have to make a separate PR for adding yourself to maintainers-list.nix. Just add an additional commit in this PR. That way, you can mark yourself as the maintainer of this package, by adding `meta.maintainers`.

**Review by mdarocha — CHANGES_REQUESTED** [review]
Consider adding the package to `pkgs/by-name` instead of the current directory - see [info](https://github.com/NixOS/nixpkgs/blob/master/pkgs/by-name/README.md)

File: pkgs/misc/universal-android-debloater/default.nix (hunk 1 above)
Comment (gesperon) [reply pkgs/misc/universal-android-debloater/default.nix]: Should be good now

File: pkgs/misc/universal-android-debloater/default.nix (hunk 2 above)
Comment (gesperon) [reply pkgs/misc/universal-android-debloater/default.nix:64]: Thank you. I just added the commit

**Comment by gesperon** [comment]
Reopened the PR... closed before by misclick.

**Review by mdarocha — APPROVED** [review]
LGTM

File: pkgs/misc/universal-android-debloater/default.nix [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+  doCheck = false;
+
+  buildFeatures = [ "iced/glow" ];
+  postInstall = ''
+    wrapProgram $out/bin/uad_gui \
+      --prefix LD_LIBRARY_PATH : ${lib.makeLibraryPath [ fontconfig freetype libglvnd xorg.libX11 xorg.libXcursor xorg.libXrandr xorg.libXi ]} \
+      --suffix PATH : ${lib.makeBinPath [ android-tools ]}
+    '';
+
+  meta = with lib; {
```
Comment (WolfangAukang) [inline pkgs/misc/universal-android-debloater/default.nix:64]: Missing a `mainProgram` on the meta

**Review by WolfangAukang — CHANGES_REQUESTED** [review]

File: pkgs/misc/universal-android-debloater/default.nix [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    fontconfig
+    freetype
+    libglvnd
+    xorg.libX11
+    xorg.libXcursor
+    xorg.libXi
+    xorg.libXrandr
+  ];
+
+  doCheck = false;
```
[the remark this answers is not in the document]
Comment (gesperon) [reply pkgs/misc/universal-android-debloater/default.nix:55]: tests would fail with:
```
...
 /build/cargo-vendor-dir/expat-sys-2.1.6/expat/tests/runtests.c:1419:40: warning: ISO C++ forbids converting a string constant to 'char*' [-Wwrite-strings]
   1419 |     TCase *tc_namespace = tcase_create("XML namespaces");
        |                                        ^~~~~~~~~~~~~~~~
  CMake Error at cmake_install.cmake:52 (file):
    file cannot create directory: /var/empty/local/lib.  Maybe need
    administrative privileges.
```

[the remark this answers is not in the document]
Comment (WolfangAukang) [reply pkgs/misc/universal-android-debloater/default.nix:55]: Yes, indicate that on a comment above. This is just to make it clear why it is disabled.

**Review by mdarocha — APPROVED** [review]
Looks fine, but looks like ofborg fails eval, you should look into fixing that

**Comment by gesperon** [comment]
> Looks fine, but looks like ofborg fails eval, you should look into fixing that

No idea what ofborg error are about.

**Comment by WolfangAukang** [comment]
Is it possible to bring the Cargo.lock file to nixpkgs? My hunch is because of this:
```
   … while realising the context of path '/nix/store/lbx1cg1fkg2j8hpp6cggr63s3xpy46a6-source/Cargo.lock'

       at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/ofborg-evaluator-3/pkgs/build-support/rust/import-cargo-lock.nix:45:10:

           44|     if lockFile != null
           45|     then builtins.readFile lockFile
             |          ^
           46|     else args.lockFileContents;

       error: path '/nix/store/kbpb2pz9n5h5vgi2kk1b867xvi93j0lh-source.drv' is not valid
```

**Comment by WolfangAukang** [comment]
Result of `nixpkgs-review pr 267729` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>universal-android-debloater</li>
  </ul>
</details>

**Review by WolfangAukang — APPROVED** [review]

**Comment by gesperon** [comment]
> Is it possible to bring the Cargo.lock file to nixpkgs? My hunch is because of this:
> 
> ```
>    … while realising the context of path '/nix/store/lbx1cg1fkg2j8hpp6cggr63s3xpy46a6-source/Cargo.lock'
> 
>        at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/ofborg-evaluator-3/pkgs/build-support/rust/import-cargo-lock.nix:45:10:
> 
>            44|     if lockFile != null
>            45|     then builtins.readFile lockFile
>              |          ^
>            46|     else args.lockFileContents;
> 
>        error: path '/nix/store/kbpb2pz9n5h5vgi2kk1b867xvi93j0lh-source.drv' is not valid
> ```

I added the Cargo.lock to nixpkgs, lets see when ofborg-eval runs.

**Comment by WolfangAukang** [comment]
@gesperon why was the PR closed? The application works fine. I guess you could ask it on [the Discourse](https://discourse.nixos.org/) to see why ofborg is failing.

**Comment by gesperon** [comment]
@WolfangAukang I wanted to share the app with the community and I don't mind maintaining it for as long as I use it. What I don't want is to spend extra time on troubleshooting random issues like ofborg-eval cryptic outputs specially when the app merges, build and runs perfectly fine.
